EnoDatabase/EnoDatabase.AttackInfo.cs

```diff
@@ -0,0 +1,87 @@
+﻿namespace EnoDatabase
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Linq;
+    using System.Text;
+    using System.Threading.Tasks;
+    using EnoCore;
+    using EnoCore.AttackInfo;
+    using EnoCore.Configuration;
+    using EnoCore.Models;
+    using Microsoft.EntityFrameworkCore;
+    using Microsoft.Extensions.Logging;
+
+    public partial class EnoDatabase : IEnoDatabase
+    {
+        public async Task<AttackInfo> GetAttackInfo(long roundId, Configuration config)
+        {
+            var teamNames = await this.context.Teams
+                .AsNoTracking()
+                .Select(t => new { t.Id, t.Address })
+                .ToDictionaryAsync(t => t.Id, t => t.Address);
+            var availableTeams = await this.context.RoundTeamServiceStatus
+                .Where(rtss => rtss.GameRoundId == roundId)
+                .GroupBy(rtss => rtss.TeamId)
+                .Select(g => new { g.Key, BestResult = g.Min(rtss => rtss.Status) })
+                .Where(ts => ts.BestResult < ServiceStatus.OFFLINE)
+                .Select(ts => ts.Key)
+                .OrderBy(ts => ts)
+                .ToArrayAsync();
+            var availableTeamNames = availableTeams.Select(id => teamNames[id] ?? id.ToString()).ToArray();
+
+            var serviceNames = await this.context.Services
+                .AsNoTracking()
+                .Select(s => new { s.Id, s.Name })
+                .ToDictionaryAsync(s => s.Id, s => s.Name);
+
+            var relevantTasks = this.context.CheckerTasks
+                .AsNoTracking()
+                .Where(ct => ct.Method == CheckerTaskMethod.putflag)
+                .Where(ct => ct.CurrentRoundId <= roundId)
+                .Where(ct => ct.CurrentRoundId > roundId - config.FlagValidityInRounds)
```

Comment:
If we have a validity of `3` rounds, shouldn't round `0` still be shown in round `3`?
right now it would be `3 - 3 =  0`, so hidden.
Just double checking.

Reply:
I'm not sure. My thinking was:

The flag is created in round `N`. It will be valid so long as `N + V >= R` with `V` being the validity in rounds and `R` being the current round ([ref](https://github.com/enowars/EnoEngine/blob/8d9cfe210ceea7a0ac1d5d09de06c1b5150aff8b/EnoDatabase/EnoDatabase.FlagSubmission.cs#L41)), so it's valid in rounds `[N; N + V]` and there are `V + 1` total rounds in which it is valid, including the one in which it is deposited.

`GenerateAttackInfo` is called at the end of round `A`. The `attack.json` generated will be displayed during round `A + 1`. Therefore it should include all flags that will be valid in round `A + 1`, which would therefore be the flags for rounds `[A + 1 - V; A + 1]`, however we cannot include flags for round `A + 1` since they will be deposited in Q1 of round `A + 1` which is only just now starting. So we want the flags for rounds `[A + 1 - V; A]`, or `]A - V; A]`.

That translates to `ct.CurrentRoundId > roundId - config.FlagValidityInRounds` and `ct.CurrentRoundId <= roundId`.

Maybe @Trolldemorted or @Savallator could weigh in.

Reply:
Yeah, since this is called at the round's end we have to ditch the oldest results